bqutil/query.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bqutil provides utils to interact with BigQuery.
+package bqutil
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/sethvargo/go-retry"
+	"google.golang.org/api/iterator"
+)
+
+// Query is the interface to execute a query.
+type Query[T any] interface {
+	// Execute runs the query and returns the result in type []*T.
+	Execute(context.Context) ([]*T, error)
```

Comment:
I'd probably make this `T` instead of `*T`. As written, if Query[*string], then you'd have to return a `**string` (which is technically valid, but super weird).

Reply:
`NewQuery[*queryResult]()`

Reply:
You're right. Not sure what I was thinking.